src/dune_rules/pkg_rules.ml

```diff
@@ -144,6 +144,47 @@ module Install_cookie = struct
   ;;
 end
 
+module Typed_env = struct
+  (* A representation of an environment where each variable can hold a
+     list of [Value.t]. Each variable will be encoded into a delimited
+     string (in the style of the PATH variable). *)
+  type t = Value.t list Env.Map.t
+
+  let parse_strings s = Bin.parse s |> List.map ~f:(fun s -> Value.String s)
+  let of_env env : t = Env.to_map env |> Env.Map.map ~f:parse_strings
+
+  let string_of_env_values values =
+    List.map values ~f:(function
+      | Value.String s -> s
+      | Dir s | Path s -> Path.to_absolute_filename s)
+    |> Bin.encode_strings
+  ;;
+
+  let to_env (t : t) = Env.Map.map t ~f:string_of_env_values |> Env.of_map
+  let get_path t = Env.Map.find t Env_path.var
+
+  (* [extend_concat_path a b] adds all variables from [b] to [a]
+     overwriting any existing values of those variables in [a] except for PATH
+     which is set to the concatenation of the PATH variables from [a] and [b]
+     with the PATH entries from [b] preceding the PATH entries from
+     [a]. If only one of the arguments contains a PATH variable then
+     its value will be the value of PATH in the result, however if
+     neither argument contains a PATH variable then PATH will be unset
+     in the result. *)
+  let extend_concat_path a b =
```

Comment:
There's a few other variables where we need this treatment - basically any variable that we expand automatically by adding a package dep. This is a good start though.

Reply:
Ah that makes sense. I'll make an issue for generalizing this to other variables. Any idea what variables might be affected or under what conditions this might fail?

Reply:
Actually @rgrinberg would you mind making an issue for this. Can you think of a problem that could be caused by dune clobering environment variables containing lists other than PATH? Possibly systems where libraries are stored in an unusual place dictated by an environment variable? I guess dune needs to establish a convention for which variables get passed through and which ones are sandboxed.

Relevant: https://github.com/ocaml/dune/pull/8237

Reply:
Isn't this the same as https://github.com/ocaml/dune/issues/10294? I've expanded the list of variables we need to relocate there.